utils/src/filter_drain.rs

```diff
@@ -0,0 +1,415 @@
+//! Backport of [`Vec::extract_if`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.extract_if) to stable Rust. Renamed to `filter_drain` to avoid naming conflict upon stabilization.
+//!
+//! See [tracking issue](https://github.com/rust-lang/rust/issues/43244).
+//!
+//! [`Original`](https://github.com/rust-lang/rust/blob/8ed95d1d9e149b5242316c91b3849c58f8320470/library/alloc/src/vec/extract_if.rs).
+
+use core::{ptr, slice};
+
+/// Extension trait that backports [`Vec::extract_if`](https://doc.rust-lang.org/stable/std/vec/struct.Vec.html#method.extract_if) which is not stable yet.
+///
+/// See [tracking issue](https://github.com/rust-lang/rust/issues/43244)
+///
+/// We call this `FilterDrain` to avoid the naming conflict with the standard library.
+pub trait FilterDrain<'a, T, F> {
+    type Item;
+    type Iter: Iterator<Item = Self::Item> + 'a;
+
+    fn filter_drain(&'a mut self, pred: F) -> Self::Iter;
+}
+
+impl<'a, T, F> FilterDrain<'a, T, F> for Vec<T>
+where
+    F: FnMut(&mut T) -> bool + 'a,
+    T: 'a,
+{
+    type Item = T;
+    type Iter = FilterDrainIter<'a, T, F>;
+
+    fn filter_drain(&'a mut self, pred: F) -> FilterDrainIter<'a, T, F> {
+        let old_len = self.len();
+
+        // Guard against us getting leaked (leak amplification)
+        unsafe {
+            self.set_len(0);
+        }
+
+        FilterDrainIter {
+            vec: self,
+            idx: 0,
+            del: 0,
+            old_len,
+            pred,
+        }
+    }
+}
+
+/// An iterator which uses a closure to determine if an element should be removed.
+#[derive(Debug)]
+#[must_use = "iterators are lazy and do nothing unless consumed"]
+pub struct FilterDrainIter<'a, T, F>
+where
+    F: FnMut(&mut T) -> bool,
+{
+    pub(super) vec: &'a mut Vec<T>,
```

Comment:
I think we do not need `pub(super)` here, because we create `FilterDrainIter` in the same module.

Reply:
In the end it doesn't matter, but you use a trait in the same module to instantiate it, while the standard library does the instantiation in the super module.